clients/src/main/java/org/apache/kafka/common/record/FileRegion.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common.record;
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.network.Send;
+import org.apache.kafka.common.network.TransferableChannel;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+
+/**
+ * Represent a file with a start offset and length, we use it to send large snapshot file
+ */
+public class FileRegion implements BaseRegion {
```

Comment:
It seems like there is similarity with how we handle records. Why do we have to duplicate this logic? cc @jsancio

Reply:
We discussed here: https://github.com/apache/kafka/pull/9553#discussion_r549053981

Reply:
I see. Not sure I'm convinced, but I'll wait to hear from @jsancio and @hachikuji.

Reply:
I agree that it is very similar to how `FileRecords` works with the difference that the file region described doesn't have to be a valid record batch. It would be interesting to see if we can unify these abstractions.

Reply:
I replaced FileRegion with FileRecords locally and it worked smoothly even the start position isn't a valid record batch, which means the requirement is a soft requirement.
So we could just use FileRecords to implement zero-copy regardless of the requirement that position should be a record batch, or we could get the max position which is a valid record batch and also smaller than request position using `OffsetIndex` to satisfy the soft requirement.